raiden/utils/filters.py

```diff
@@ -89,3 +97,42 @@ def decode_event(abi: Dict, log: Dict):
     }
     event_abi = topic_to_event_abi[event_id]
     return get_event_data(event_abi, log)
+
+
+class StatelessFilter(LogFilter):
+    _last_block: int = -1
+
+    def __init__(self, web3: Web3, filter_params: dict):
+        super().__init__(web3, filter_id=None)
+        self.filter_params = filter_params
+        self._lock = Semaphore()
+
+    def get_new_entries(self):
+        with self._lock:
+            filter_params = self.filter_params.copy()
+            filter_params['fromBlock'] = max(
+                filter_params.get('fromBlock', 0),
+                self._last_block + 1,
+            )
+            if self.filter_params.get('toBlock') in ('latest', 'pending'):
+                filter_params['toBlock'] = self.web3.eth.blockNumber
+            self._update_last_block(filter_params.get('toBlock', -1))
```

Comment:
```python
# This logic contains a race condition. It's possible that after `web.eth.blockNumber` and before
# `web3.eth.getLogs` a new block is mined. This is okay because any new logs on this new block
# will be fetched on the next call to `get_new_entries`
if self.filter_params.get('toBlock') in ('latest', 'pending'):
     block_number = self.web3.eth.blockNumber
     self._last_block = block_number
     filter_params['toBlock'] = block_number
```
Is easier to understand IMO

Reply:
One of the arguments for this change is to reduce the number of RPC calls. With installed filters we always have two additional RPC calls (one to install and another to uninstall the filters), so in total we have `2 + n` RPC calls, were `n` is the number of times the node queries the filter (ideally once per block).

This implementation is removing the constant `2`, by not using installed filters, however it's querying the block number for each call to `get_new_entries`, which effectly make it `2n` RPC calls. So in number of RPC calls this is *worse* then our previous implementation, however I think this can be fixed by passing the `block_number` as an argument to `get_new_entries`

Reply:
I'd rather not pass a new parameter to `get_new_entries`, as I'd prefer to keep it compatible with the web3's `Filter` interface. But as it's only instantiated by our `JSONRPCClient`, which also is the interface through which the `block_number` is fetched, I can cache it there and get the cached value from the filters.

Reply:
> I'd rather not pass a new parameter to get_new_entries, as I'd prefer to keep it compatible with the web3's Filter interface. 

Why? Is it keeping compatibility with a third party dependency more important than fixing race conditions?

Reply:
The race condition **was** removed when I fixed the `toBlock` to the latest known. About the uneeded double request, I fixed it by actually getting the same value, but from a different path: caching it in the JSONRPCClient and fetching this cached block number from the filter (which would be the same as the number the client would have access)

Reply:
> The race condition was removed when I fixed the toBlock to the latest known

Maybe we are not talking about the same thing, the race I'm referring to cannot be removed, we can not stop the blockchain. It's always possible to query for the current block number just before a new block is mined.

What we can do is have a consistent view of the known latest block inside the node, for that to work we need to have a single place were this update can happen (i.e. the [alarm task](https://github.com/raiden-network/raiden/blob/6444580a4208a32afcebc5415c4c34db671967c2/raiden/tasks.py#L77-L100)), the rest of the code base should use [that value as the "source of truth"](https://github.com/raiden-network/raiden/blob/master/raiden/raiden_service.py#L326-L327), and this needs to be protected [by a lock to serialize new updates](https://github.com/raiden-network/raiden/blob/master/raiden/raiden_service.py#L375).

Note: The existing code *is also* susceptible to races on the filter, which needs to be fixed.

Reply:
@hackaugusto Ok, I'd rather find a better solution in the future, but went with your suggestion now.